File: app/src/main/java/com/github/tvbox/osc/base/BaseActivity.java
package com.github.tvbox.osc.base;

import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.Looper;
import android.util.DisplayMetrics;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.PermissionChecker;

import com.github.tvbox.osc.R;
import com.github.tvbox.osc.callback.EmptyCallback;
import com.github.tvbox.osc.callback.LoadingCallback;
import com.github.tvbox.osc.util.AppManager;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.MaterialToolbar;
import com.kingja.loadsir.callback.Callback;
import com.kingja.loadsir.core.LoadService;
import com.kingja.loadsir.core.LoadSir;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import me.jessyan.autosize.AutoSizeCompat;
import me.jessyan.autosize.internal.CustomAdapt;
import xyz.doikki.videoplayer.util.CutoutUtil;

/**
 * @author pj567
 * @date :2020/12/17
 * @description:
 */
public abstract class BaseActivity extends AppCompatActivity implements CustomAdapt {
    private static float screenRatio = -100.0f;
    protected Context mContext;
    private LoadService<?> mLoadService;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        try {
            if (screenRatio < 0) {
                DisplayMetrics dm = new DisplayMetrics();
                getWindowManager().getDefaultDisplay().getMetrics(dm);
                int screenWidth = dm.widthPixels;
                int screenHeight = dm.heightPixels;
                screenRatio = (float) Math.max(screenWidth, screenHeight) / (float) Math.min(screenWidth, screenHeight);
            }
        } catch (Throwable th) {
            th.printStackTrace();
        }
        super.onCreate(savedInstanceState);
        setContentView(getLayoutResID());
        mContext = this;
        CutoutUtil.adaptCutoutAboveAndroidP(mContext, true);
        AppManager.getInstance().addActivity(this);
        setupAppBarTransparency();
        init();
    }

    /**
     * 自动为 AppBarLayout 设置透明度渐变效果
     * 当标题栏向上收起时，透明度会逐渐降低，避免与状态栏内容重叠
     */
    private void setupAppBarTransparency() {
        try {
            AppBarLayout appBarLayout = findViewById(R.id.appBarLayout);
            MaterialToolbar appBar = findViewById(R.id.appBar);

            if (appBarLayout != null && appBar != null) {
                appBarLayout.addOnOffsetChangedListener(
                        (appBarLayout1, verticalOffset) -> {
                            int totalScrollRange = appBarLayout1.getTotalScrollRange();
                            if (totalScrollRange == 0) return;

                            // 计算滚动进度（0.0 = 完全展开，1.0 = 完全收起）
                            float scrollProgress = Math.abs(verticalOffset) / (float) totalScrollRange;

                            // 设置标题栏透明度（收起时变透明）
                            appBar.setAlpha(1.0f - scrollProgress);
                        }
                );
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public Resources getResources() {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            AutoSizeCompat.autoConvertDensityOfCustomAdapt(super.getResources(), this);
        }
        return super.getResources();
    }

    public boolean hasPermission(String permission) {
        boolean has = true;
        try {
            has = PermissionChecker.checkSelfPermission(this, permission) == PermissionChecker.PERMISSION_GRANTED;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return has;
    }

    protected abstract int getLayoutResID();

    protected abstract void init();

    protected void setLoadSir(View view) {
        if (mLoadService == null) {
            mLoadService = LoadSir.getDefault().register(view, (Callback.OnReloadListener) v -> {
            });
        }
    }

    protected void showLoading() {
        if (mLoadService != null) {
            mLoadService.showCallback(LoadingCallback.class);
        }
    }

    protected boolean isLoading() {
        if (mLoadService != null && mLoadService.getCurrentCallback() != null) {
            return mLoadService.getCurrentCallback().equals(LoadingCallback.class);
        }
        return false;
    }

    protected void showEmpty() {
        if (null != mLoadService) {
            mLoadService.showCallback(EmptyCallback.class);
        }
    }

    protected void showSuccess() {
        if (null != mLoadService) {
            mLoadService.showSuccess();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        AppManager.getInstance().finishActivity(this);
    }

    public void jumpActivity(Class<? extends BaseActivity> clazz) {
        Intent intent = new Intent(mContext, clazz);
        startActivity(intent);
    }

    public void jumpActivity(Class<? extends BaseActivity> clazz, Bundle bundle) {
        Intent intent = new Intent(mContext, clazz);
        intent.putExtras(bundle);
        startActivity(intent);
    }

    protected String getAssetText(String fileName) {
        StringBuilder stringBuilder = new StringBuilder();
        try {
            AssetManager assets = getAssets();
            BufferedReader bf = new BufferedReader(new InputStreamReader(assets.open(fileName)));
            String line;
            while ((line = bf.readLine()) != null) {
                stringBuilder.append(line);
            }
            return stringBuilder.toString();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return "";
    }

    @Override
    public float getSizeInDp() {
        return isBaseOnWidth() ? 1280 : 720;
    }

    @Override
    public boolean isBaseOnWidth() {
        return !(screenRatio >= 4.0f);
    }
}

File: app/src/main/java/com/github/tvbox/osc/base/BaseActivity.kt
package com.github.tvbox.osc.base

import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle
import android.os.Looper
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.PermissionChecker
import com.github.tvbox.osc.R
import com.github.tvbox.osc.callback.EmptyCallback
import com.github.tvbox.osc.callback.LoadingCallback
import com.github.tvbox.osc.util.AppManager
import com.kingja.loadsir.core.LoadService
import com.kingja.loadsir.core.LoadSir
import me.jessyan.autosize.AutoSizeCompat
import me.jessyan.autosize.internal.CustomAdapt
import kotlin.math.max
import kotlin.math.min

/**
 * @author pj567
 * @date :2020/12/17
 */
abstract class BaseActivity : AppCompatActivity(), CustomAdapt {
	protected lateinit var mContext: Context
	protected open val layoutResID: Int = 0
	private var mLoadService: LoadService<*>? = null

	fun jumpActivity(clazz: Class<out BaseActivity>) {
		val intent = Intent(mContext, clazz)
		startActivity(intent)
	}

	fun jumpActivity(clazz: Class<out BaseActivity>, bundle: Bundle) {
		val intent = Intent(mContext, clazz)
		intent.putExtras(bundle)
		startActivity(intent)
	}

	fun hasPermission(permission: String): Boolean {
		var has = true
		try {
			has = PermissionChecker.checkSelfPermission(this, permission) == PermissionChecker.PERMISSION_GRANTED
		} catch (e: Exception) {
			e.printStackTrace()
		}
		return has
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		try {
			if (screenRatio < 0) {
				val dm = resources.displayMetrics
				val screenWidth = dm.widthPixels
				val screenHeight = dm.heightPixels
				screenRatio = max(screenWidth, screenHeight).toFloat() / min(screenWidth, screenHeight).toFloat()
			}
		} catch (th: Throwable) {
			th.printStackTrace()
		}
		super.onCreate(savedInstanceState)
		if (layoutResID != 0) setContentView(this.layoutResID)
		mContext = this
		AppManager.instance.addActivity(this)
		init()
	}

	override fun onResume() {
		super.onResume()
	}

	override fun getResources(): Resources {
		if (Looper.myLooper() == Looper.getMainLooper()) {
			AutoSizeCompat.autoConvertDensityOfCustomAdapt(super.getResources(), this)
		}
		return super.getResources()
	}

	override fun getSizeInDp(): Float {
		return (if (isBaseOnWidth) 1280 else 720).toFloat()
	}

	override fun isBaseOnWidth(): Boolean {
		return !(screenRatio >= 4.0f)
	}

	override fun onDestroy() {
		super.onDestroy()
		AppManager.instance.finishActivity(this)
	}

	protected abstract fun init()

	protected fun setLoadSir(view: View?) {
		if (mLoadService == null) {
			mLoadService = LoadSir.getDefault().register(view) { v: View? -> }
		}
	}

	protected fun showLoading() {
		mLoadService?.showCallback(LoadingCallback::class.java)
	}

	protected fun showEmpty() {
		mLoadService?.showCallback(EmptyCallback::class.java)
	}

	protected fun showSuccess() {
		mLoadService?.showSuccess()
	}

	/**
	 * 自动为 AppBarLayout 设置透明度渐变效果
	 * 当标题栏向上收起时，透明度会逐渐降低，避免与状态栏内容重叠
	 */
	companion object {
		private var screenRatio = -100.0f
	}
}